### PR TITLE
[Spells] Implemented SPA 476 SE_Weapons_Stance and Live-like AA Enable/Disable Toggle

### PR DESCRIPTION
**Pull request by KayenEQ**
UPDATED ON 8/3/21 - Toggle system now implemented very close to live using new spell effect SE_Buy_AA_Rank (472). This
spell effect must be placed on the spell you place in the Disabled rank. Again, the SQL is updated on bottom for testing. Two spells SQL used in the AA are also added, can insert into default PEQ spell table without any issues.

**Please note! The new toggle system can be used to Enable or Disable ANY type of passive AA effect you want. You obviously
need to customize your AA to do this though. Follow the example here.**

------------------------------------------------------------------------------------------------------------------------------------

Implementation testing for SPA 476 SE_Weapon_Stance

Allows for automatic application of specific buffs depending on if you have a 2hander, shield or duel wield equipped. The buff will change depending on what is equipped.
Base1: spell id base2 weapon type (0=2handed, 1=Shield, 2=dual wield)

Can be applied as a Spell Buff, Item worn effect, AA passive ability
You can have 1 of each of these types active at a time. Resulting in up to 3 buffs per weapon type.

Note: At the bottom you will find SQL you can run to test all the examples given here. These SQL are ONLY
FOR TESTING PURPOSES.

Example for Spell Buff.
 
I edit the Charisma Buff - Alluring Aura to contain the weapon stance effect. 
I want it to buff me with a shaman 'Strength buff' if I have a 2handed weapon, 'Agility buff' if I have a Shield, and 'Dexterity buff' if Dual wield
So I set effectid2,3,4 are all 476
I chose to use Spirit of Strength, Spirit of Cat and Spirit of Monkey as my spells
then Set effect_base_value1,2,3 to spellids 147 (Spirit of Strength),148(Spirit of Cat),146 Spirit of Monkey) respectively
then set effect_limit_value1,2,3 to 0(2H), 1(Shield), 2(Duel wield) respectively.

Now when I cast Alluring Aura (SPELL ID 150) it will give me the charisma buff, AND depending on which weapons/shield I have it will give me a different stat buff.
If you want an upgraded version with improved stat buffs, try casting Charisma (SPELL ID 156)

Example for item buff.
Run script bellow to add to your database two test items. Weaponstance Earring Rank 1 (150000) Weaponstance Earring Rank 2 (150002)
 
I edit the HP buff - Skin like Wood to contain the weapon stance effect. Place Skin like Wood as a worn effect on 'Weaponstance Earring Rank 1'
I want it to buff me with Fire resist if I have a 2handed weapon, Cold resist if I have a Shield, and Magic resist if Dualweild
So I set effectid2,3,4 are all 476
I chose to use endure fire, cold and magic as my spells
then Set effect_base_value1,2,3 to spellids 224 (endure fire),225(endure cold),228(endure magic) respectively
then set effect_limit_value1,2,3 to 0(2H), 1(Shield), 2(Duelweild) respectively.

Simply equip the item to get the new effects.
If you want to see upgraded version with better resist buffs. Equip Weaponstance Earring Rank 2 (150001).
Note, you should always get the best buff even if both are equiped at the same time.

Example for AA passive effects. (Live uses this only in AA passive effects)
Run SQL scripts below to generate a new AA, ability ID will be 1600 called 'Weapon Stance TEST ONLY'
In game will appear for RANGERS, under class tab, with name 'BETA ONLY Destructive Force Test'
AA Toggle effect is coded to resemble lives. When you buy the AA it will give you a hotkey called 'Reactive'(best I could find,
can always make better custom hotkeys)

-Where first rank you buy is the DISABLED version, in game it will look like BETA ONLY Destructive Force Test' 1/4

-When you click the hotkey it will buy the next rank and this is the ENABLED version and you should get buff effects.
In gave it will look like 'BETA ONLY Destructive Force Test' 2/4

If you click the hotkey now, it will disable the ability and put your rank back to 'BETA ONLY Destructive Force Test' 1/4.

AA will work as follows
I want it to buff me with Regen if I have a 2handed weapon, Breeze resist if I have a Shield, and Celerity if Dualweild
So I added in 'aa_rank_effets slot 1,2,3 for aa_rank_id 1345 and set 'effect_id' for each slot to 476
then Set base1 for slots 1,2,3 to spellids 144 (regen),697(breeze),171(celerity) respectively
then set base2 1, 2,3 to 0(2H), 1(Shield), 2(Duelweild) respectively.

You can upgrade the focus buy purchasing a new rank when your at  'BETA ONLY Destructive Force Test' 2/4. The pattern
will then continue with 3/4 and 4/4 ect.

---------------------------------
DETAILS ON HOW TO SET UP THE AA - again full example SQL at bottom (This information is in source code)

		Certain AA, like Weapon Stance line use a special toggle Hotkey to enable or disable the AA's passive abilities.
		This is occurs by doing the following. Each 'rank' of Weapon Stance is actually 2 actual ranks.
		First rank is always the Disabled version which cost X amount of AA. Second rank is the Enabled version which cost 0 AA.
		When you buy the first rank, you make a hotkey that on live say 'Weapon Stance Disabled', if you clik that it then BUYS the
		next rank of AA (cost 0) which switches the hotkey to 'Enabled Weapon Stance' and you are given the passive buff effects.
		If you click the Enabled hotkey, it causes you to lose an AA rank and once again be disabled. Thus, you are switching between
		two AA ranks. Thefore when creating an AA using this ability, you need generate both ranks. Follow the same pattern for additional ranks.

		IMPORTANT! The toggle system can be used to Enable or Disable ANY passive AA. You just need to follow the instructions on how to create it.
		Example: Enable or Disable a buff that gives a large hate modifier. Play may Enable when tanking and Disable when DPS ect.

		Note: On live the Enabled rank is shown having a Charge of 1, while Disabled rank has no charges. Our current code doesn't support that. Do not use charges.
		Note: Live uses a spell 'Disable Ability' ID 46164 to trigger a script to do the AA rank changes. At present time it is not coded to require that, any spell id works.
		Note: Discovered a bug on ROF2, where when you buy first rank of an AA with a hotkey, it will always display the title of the second rank in the database. Be aware. No easy fix.

		Dev Note(Kayen 8/1/21): The system as set up is very similar to live, with exception that live gives the Enabled rank 1 Charge. The code here emulates what happens when a
		charge would be expended.

		Instructions for how to make the AA - assuming a basic level of knowledge of how AA's work.
		- aa_abilities table : Create new ability with a hotkey, type 3, zero charges
		- aa_ranks table :  [Disabled rank] First rank, should have a cost > 0 (this is what you buy), Set hotkeys, MUST SET A SPELL CONTAINING EFFECT SE_Buy_AA_Rank(SPA 472), set a short recast timer.
							[Enabled rank] Second rank, should have a cost = 0, Set hotkeys, Set any valid spell ID you want (it has to exist but does nothing), set a short recast timer.
							*Recommend if doing custom, just make the hotkey titled 'Toggle <Ability Name>' and use for both.

		- aa_rank_effects table : [Disabled rank] No data needed in the aa_ranks_effect table
								  [Enabled rank]  Second rank set effect_id = 457 (weapon stance), slot 1,2,3, base1= spell triggers, base= weapon type (0=2H,1=SH,2=DW), for slot 1,2,3

			Example SQL			-Disabled
								DO NOT ADD any data to the aa_rank_effects for this rank_id

								-Enabled
								INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20003, 1, 476, 145,0);
								INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20003, 2, 476, 174,1);
								INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20003, 3, 476, 172,2);

		Warning: If you want to design an AA that only uses one weapon type to trigger, like will only apply buff if Shield. Do not include data for other types. Never have a base value=0
		in the Enabled rank.
---------------------------------
WARNING BEFORE YOU RUN THESE SQLs
- Make a backup COPY of your 'spells_new', 'items', and 'aa_rank_effects' table. You will want to restore them when done.
--------------------------------

----------------------------------------------
Best way to test: 
-Create a new RANGER. 
-Summon a shield, two handed weapon and two swords. 
-Try casting example spell. Try equipping example items. Try learning example AA.
---------------------------------------------------------------

```
-- Spell
UPDATE spells_new SET rank = 1 WHERE id = 147;
UPDATE spells_new SET rank = 1 WHERE id = 148;
UPDATE spells_new SET rank = 1 WHERE id = 146;

UPDATE spells_new SET rank = 2 WHERE id = 159;
UPDATE spells_new SET rank = 2 WHERE id = 160;
UPDATE spells_new SET rank = 2 WHERE id = 157;

-- AA
UPDATE spells_new SET rank = 1 WHERE id = 144;
UPDATE spells_new SET rank = 1 WHERE id = 697;
UPDATE spells_new SET rank = 1 WHERE id = 171;

UPDATE spells_new SET rank = 2 WHERE id = 145;
UPDATE spells_new SET rank = 2 WHERE id = 174;
UPDATE spells_new SET rank = 2 WHERE id = 172;

-- Item
UPDATE spells_new SET rank = 1 WHERE id = 224;
UPDATE spells_new SET rank = 1 WHERE id = 225;
UPDATE spells_new SET rank = 1 WHERE id = 228;

UPDATE spells_new SET rank = 2 WHERE id = 60;
UPDATE spells_new SET rank = 2 WHERE id = 61;
UPDATE spells_new SET rank = 2 WHERE id = 64;

-- Spell
UPDATE spells_new SET effectid2 = 476, effect_base_value2 =147, effect_limit_value2 =0  WHERE id =150;
UPDATE spells_new SET effectid3 = 476, effect_base_value3 =148, effect_limit_value3 =1  WHERE id =150;
UPDATE spells_new SET effectid4 = 476, effect_base_value4 =146, effect_limit_value4 =2  WHERE id =150;

UPDATE spells_new SET effectid2 = 476, effect_base_value2 =159, effect_limit_value2 =0  WHERE id =156;
UPDATE spells_new SET effectid3 = 476, effect_base_value3 =160, effect_limit_value3 =1  WHERE id =156;
UPDATE spells_new SET effectid4 = 476, effect_base_value4 =157, effect_limit_value4 =2  WHERE id =156;


-- Item
UPDATE spells_new SET effectid2 = 476, effect_base_value2 =224, effect_limit_value2 =0  WHERE id =26;
UPDATE spells_new SET effectid3 = 476, effect_base_value3 =225, effect_limit_value3 =1  WHERE id =26;
UPDATE spells_new SET effectid4 = 476, effect_base_value4 =228, effect_limit_value4 =2  WHERE id =26;

UPDATE spells_new SET effectid2 = 476, effect_base_value2 =60, effect_limit_value2 =0  WHERE id =423;
UPDATE spells_new SET effectid3 = 476, effect_base_value3 =61, effect_limit_value3 =1  WHERE id =423;
UPDATE spells_new SET effectid4 = 476, effect_base_value4 =64, effect_limit_value4 =2  WHERE id =423;

--- Test Items
INSERT INTO `items` VALUES ('150000', '0', 'Weaponstance Earring Rank 1', '0', '2', '0', '-10', '0', '5', '0', '5', '5', '0', '0', '7', '1', '0', '1', '0', '1', '0', '1', '0', '1', '0', '1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '', '0', '65535', '4278190080', '0', '0', '0', '78955', '5', '0', '0', '0', '0', '0', '0', '-8', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '', '-1', '5', '0', '0', '0', '0', '0', '756', 'IT63', '0', '29', '0', '0', '0', '0', 'Fingerbone Hoop', '-1', '1', '0', '0', '0', '0', '0', '0', '5', '1', '1', '0', '-8', '0', '65535', '0', '0', '0', '0', '7', '0', '0', '-1', '0', '18', '-1', '0', '0', '0', '0', '0', '852', '1', '0', '1', '0', '0', '0', '0', '0', '0', '852', '0', '0', '0', '0', '2015-12-11 21:00:53', 'Adjusted difference on CR/DR between 13th and Lucy', '0', '0', '0', '0', '1', '0', '0', '', '0', '-1', '0', '0', '0', '0', '26', '2', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '0', '0', '0', '2021-07-29 18:12:52', '2016-01-06 06:03:22', '1', '13THFLOOR', '0', '', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '0', '00000000000000000000', '0', '', '-1', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '', '0', '0', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '0', '', '', '', '', '', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '2015-12-11 21:00:53', '0', '70', '0', '0', '0', '-1', '0', '0', '0', '0', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '-256', '255', '0', '0', '0', '0', '0', '0', '0', '0');

INSERT INTO `items` VALUES ('150001', '0', 'Weaponstance Earring Rank 2', '0', '2', '0', '-10', '0', '5', '0', '5', '5', '0', '0', '7', '1', '0', '1', '0', '1', '0', '1', '0', '1', '0', '1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '', '0', '65535', '4278190080', '0', '0', '0', '78955', '5', '0', '0', '0', '0', '0', '0', '-8', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '', '-1', '5', '0', '0', '0', '0', '0', '501', 'IT63', '0', '29', '0', '0', '0', '0', 'Fingerbone Hoop', '-1', '1', '0', '0', '0', '0', '0', '0', '5', '1', '1', '0', '-8', '0', '65535', '0', '0', '0', '0', '7', '0', '0', '-1', '0', '18', '-1', '0', '0', '0', '0', '0', '852', '1', '0', '1', '0', '0', '0', '0', '0', '0', '852', '0', '0', '0', '0', '2015-12-11 21:00:53', 'Adjusted difference on CR/DR between 13th and Lucy', '0', '0', '0', '0', '1', '0', '0', '', '0', '-1', '0', '0', '0', '0', '423', '2', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '0', '0', '0', '2021-07-29 18:12:52', '2016-01-06 06:03:22', '1', '13THFLOOR', '0', '', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '0', '00000000000000000000', '0', '', '-1', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '', '0', '0', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '0', '', '', '', '', '', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '2015-12-11 21:00:53', '0', '70', '0', '0', '0', '-1', '0', '0', '0', '0', '0', '0', '0', '0', '', '-1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '-256', '255', '0', '0', '0', '0', '0', '0', '0', '0');

-- AA
INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20001, 1, 476, 144,0);
INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20001, 2, 476, 697,1);
INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20001, 3, 476, 171,2);

INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20003, 1, 476, 145,0);
INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20003, 2, 476, 174,1);
INSERT INTO aa_rank_effects (rank_id, slot, effect_id, base1, base2) VALUES (20003, 3, 476, 172,2);

INSERT INTO `aa_ranks` VALUES ('20000', '15496', '6290', '6600', '12884', '5', '1', '42284, '6', '1', '8', '-1', '20001');
INSERT INTO `aa_ranks` VALUES ('20001', '15496', '6290', '6600', '12884', '0', '1', '42285', '6', '1', '8', '20000', '20002');
INSERT INTO `aa_ranks` VALUES ('20002', '15496', '6290', '6600', '12884', '5', '2', '42284', '6', '1', '8', '20001', '20003');
INSERT INTO `aa_ranks` VALUES ('20003', '15496', '6290', '6600', '12884', '0', '2', '42285', '6', '1', '8', '20003', '-1');

INSERT INTO `aa_ability` VALUES ('16000', 'Weapon Stance TEST ONLY', '-1', '8', '65535', '127', '131071', '0', '3', '0', '0', '20000', '1', '0');

-- Required spell to make AA work.
INSERT INTO `spells_new` VALUES ('42284', 'Enable Ability', 'PLAYER_1', '', '', '', '', '', '', '100', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '2501', '2054', '-1', '-1', '-1', '-1', '1', '1', '1', '1', '-1', '-1', '-1', '-1', '100', '100', '100', '100', '100', '100', '100', '100', '100', '100', '100', '100', '0', '1', '0', '0', '472', '254', '254', '254', '254', '254', '254', '254', '254', '254', '254', '254', '6', '25', '5', '-1', '0', '0', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '43', '0', '0', '7', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '100', '0', '119', '208', '0', '0', '0', '1', '0', '0', '0', '0', '0', '222', '42', '94', '0', '0', '0', '0', '0', '100', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '5', '101', '14', '114', '5', '3', '0', '0', '0', '0', '0', '50', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '1', '1', '0', '0', '0', '0', '0', '-1', '0', '0', '0', '1', '0', '0', '1', '1', '1', '0', '-1', '0', '0', '0', '1', '39', '-1', '0', '1', '0', '0', '1', '0', '1', '0', '0', '0', '0', '0', '0');

INSERT INTO `spells_new` VALUES ('42285', 'Disable Ability', 'PLAYER_1', '', '', '', '', '', '', '100', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '2501', '2054', '-1', '-1', '-1', '-1', '1', '1', '1', '1', '-1', '-1', '-1', '-1', '100', '100', '100', '100', '100', '100', '100', '100', '100', '100', '100', '100', '0', '1', '0', '0', '254', '254', '254', '254', '254', '254', '254', '254', '254', '254', '254', '254', '6', '25', '5', '-1', '0', '0', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '255', '43', '0', '0', '7', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '100', '0', '119', '208', '0', '0', '0', '1', '0', '0', '0', '0', '0', '222', '42', '94', '0', '0', '0', '0', '0', '100', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '5', '101', '14', '114', '5', '3', '0', '0', '0', '0', '0', '50', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '1', '1', '0', '0', '0', '0', '0', '-1', '0', '0', '0', '1', '0', '0', '1', '1', '1', '0', '-1', '0', '0', '0', '1', '39', '-1', '0', '1', '0', '0', '1', '0', '1', '0', '0', '0', '0', '0', '0');


```

### CODE REVIEW

File: zone/aa.cpp [hunk 1]
```diff
@@ -1796,3 +1797,43 @@ bool Mob::CheckAATimer(int timer)
...
+
+void Client::TogglePassiveAA(const AA::Rank &rank, int spell_id)
+{
+	/*
+	Live passive AA effects that can be toggled by hotkey include in the passive effect a trigger on cast of spell "Disable Ability" id 46164
+	The spell "Disable Ability" contains no actual SPA id or data, thus it must be hardcoded to do this effect when triggered.
+	Since this spell does not exist on our current database (7/29/21) and we don't have innate AA yet who would naturally use it. Will hold off on implementation using it.
+	*/
+
+	if (spell_id == 46164) {
```
Comment (Akkadius) [inline zone/aa.cpp:1809]: Can we create a constant for this spell and reference it please

File: zone/aa.cpp [hunk 2]
```diff
@@ -1796,3 +1797,43 @@ bool Mob::CheckAATimer(int timer)
...
+		//Disable *Not implemented yet. If we import AA data that supports it at someone will be made aware by the message.
+		Message(Chat::Red, "Use of Disably Ability through spell id 46164 is not implemented at this time, contact Kayen who will likely be in a retirement home at this time.");
+		return;
+	}
+
+	//Can add any specific use cases below.
+
+	int effect = 0;
+
```
Comment (Akkadius) [inline zone/aa.cpp:1818]: Extra newline

File: zone/aa.cpp [hunk 3]
```diff
@@ -1796,3 +1797,43 @@ bool Mob::CheckAATimer(int timer)
...
+	for (const auto &e : rank.effects) {
+		effect = e.effect_id;
+
+		switch (effect) {
+
```
Comment (Akkadius) [inline zone/aa.cpp:1823]: Newlines

File: zone/bonuses.cpp [hunk 4]
```diff
@@ -173,6 +174,9 @@ void Client::CalcItemBonuses(StatBonuses* newbon) {
 		}
 		else if (i == EQ::invslot::slotPrimary && (item && (item->ItemType == EQ::item::ItemType2HSlash || item->ItemType == EQ::item::ItemType2HPiercing)))
 			SetTwoHanderEquipped(true);
+		
+		if (CanThisClassDualWield())
```
Comment (Akkadius) [inline zone/bonuses.cpp:178]: Close these out bracket wise please

File: zone/bonuses.cpp [hunk 5]
```diff
@@ -1568,6 +1574,25 @@ void Mob::ApplyAABonuses(const AA::Rank &rank, StatBonuses *newbon)
 			newbon->Pet_Add_Atk += base1;
 			break;
 
+		case SE_Weapon_Stance: 
+		{
+			if (IsValidSpell(base1)) { //base1 is the spell_id of buff
+				if (base2 < 3){ //0=2H, 1=Shield, 2=DW
```
Comment (Akkadius) [inline zone/bonuses.cpp:1580]: Could benefit from a constant or enum here

File: zone/bonuses.cpp [hunk 6]
```diff
@@ -1568,6 +1574,25 @@ void Mob::ApplyAABonuses(const AA::Rank &rank, StatBonuses *newbon)
 			newbon->Pet_Add_Atk += base1;
 			break;
 
+		case SE_Weapon_Stance: 
+		{
+			if (IsValidSpell(base1)) { //base1 is the spell_id of buff
```
Comment (Akkadius) [inline zone/bonuses.cpp:1579]: Whenever there is a comment explaining things, there might be an opportunity to codeify this

Eg `int32 spell_id = base1;`

File: zone/common.h [hunk 7]
```diff
@@ -544,7 +544,7 @@ struct StatBonuses {
 	int32	DS_Mitigation_Percentage;			// base = percent amt of DS mitigation. Negative value to reduce
 	int32   Pet_Crit_Melee_Damage_Pct_Owner;	// base = percent mod for pet critcal damage from owner
 	int32	Pet_Add_Atk;						// base = Pet ATK bonus from owner
-
+	int32   WeaponStance[3];					// base = trigger spell id, base2 = 0 is 2h, 1 is shield, 2 is dual wield, [0]spid 2h, [1]spid shield, [2]spid DW
```
Comment (Akkadius) [inline zone/common.h:547]: `snake_case`

File: zone/client.cpp [hunk 8]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+	Toggling ON or OFF
+	- From spells, just remove the Primary buff that contains the WeaponStance effect in it.
+	- For items with worn effect, unequip the item.
+	- For AA abilities, need to make sure any AA made with this also has a Hotkey. The hotkey will serve as toggle on/off, checked in TogglePassiveAA in aa.cpp.
+	Rank
+	- Most important for AA, but if you have more than one of WeaponStance effect for a given type, the spell trigger buff will apply whatever has the highest
+	rank value from the spells table. AA's on live for this effect naturally do this. Be awere of this if making custom spells/worn effects/AA.
+	*/
+
+	if (!IsWeaponStanceEnabled())
```
Comment (Akkadius) [inline zone/client.cpp:10267]: Close out

File: zone/client.cpp [hunk 9]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+		return;
+	
+	bool enabled = false;
+	bool item_bonus_exists = false;
+	bool aa_bonus_exists = false;
+
+	if (weaponstance.spellbonus_enabled) {
+
+		if (spellbonuses.WeaponStance[0] || spellbonuses.WeaponStance[1] || spellbonuses.WeaponStance[2]) {
```
Comment (Akkadius) [inline zone/client.cpp:10276]: Replace array accessors with enums

File: zone/client.cpp [hunk 10]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
 	if (this->Admin() != newStatus)
 		database.UpdateGMStatus(this->AccountID(), newStatus);
 }
+
+void Client::ApplyWeaponsStance()
+{
+	/*
+	If you have a weapons stance bonus from at least one bonus type, each time you change weapons this function will ensure the correct
```
Comment (Akkadius) [inline zone/client.cpp:10252]: Tab in once

File: zone/client.cpp [hunk 11]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+	rank value from the spells table. AA's on live for this effect naturally do this. Be awere of this if making custom spells/worn effects/AA.
+	*/
+
+	if (!IsWeaponStanceEnabled())
+		return;
+	
+	bool enabled = false;
+	bool item_bonus_exists = false;
+	bool aa_bonus_exists = false;
+
```
Comment (Akkadius) [inline zone/client.cpp:10273]: newline

File: zone/client.cpp [hunk 12]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+				}
+				weaponstance.spellbonus_buff_spell_id = spellbonuses.WeaponStance[2];
+			}
+		}
+	}
+
+	//Spellbonus effect removal is checked in BuffFadeBySlot(int slot, bool iRecalcBonuses) in spell_effects.cpp when the buff is clicked off or fades.
+
+	if (weaponstance.itembonus_enabled) {
+
```
Comment (Akkadius) [inline zone/client.cpp:10320]: newline

File: zone/client.cpp [hunk 13]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+		if (itembonuses.WeaponStance[0] || itembonuses.WeaponStance[1] || itembonuses.WeaponStance[2]) {
+
+			enabled = true;
+			item_bonus_exists = true;
+
```
Comment (Akkadius) [inline zone/client.cpp:10325]: extra newlines

File: zone/client.cpp [hunk 14]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+		if (spellbonuses.WeaponStance[0] || spellbonuses.WeaponStance[1] || spellbonuses.WeaponStance[2]) {
+
+			enabled = true;
+
+			//Check if no longer has correct combination of weapon type and buff, if so remove buff.
+			if (!HasTwoHanderEquipped() && IsBuffSpell(spellbonuses.WeaponStance[0]) && FindBuff(spellbonuses.WeaponStance[0])) {
+				BuffFadeBySpellID(spellbonuses.WeaponStance[0]);
+			}
+
```
Comment (Akkadius) [inline zone/client.cpp:10284]: extra newline between if blocks

File: zone/client.cpp [hunk 15]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+		if (itembonuses.WeaponStance[0] || itembonuses.WeaponStance[1] || itembonuses.WeaponStance[2]) {
+
+			enabled = true;
+			item_bonus_exists = true;
+
+
+			//Edge case check if have multiple items with WeaponStance worn effect. Make sure correct buffs are applied if items are removed but others left on.
+			if (weaponstance.itembonus_buff_spell_id){
+
```
Comment (Akkadius) [inline zone/client.cpp:10329]: newline

File: zone/client.cpp [hunk 16]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+					weaponstance.itembonus_buff_spell_id = 0; //Need to zero this before we fade to prevent any recursive loops.
+					BuffFadeBySpellID(fade_spell);
+				}
+			}
+
+			//Check if no longer has correct combination of weapon type and buff, if so remove buff.
+			if (!HasTwoHanderEquipped() && IsBuffSpell(itembonuses.WeaponStance[0]) && FindBuff(itembonuses.WeaponStance[0])) {
+				BuffFadeBySpellID(itembonuses.WeaponStance[0]);
+			}
+
```
Comment (Akkadius) [inline zone/client.cpp:10346]: newlines between if blocks

File: zone/client.cpp [hunk 17]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+			//If you have correct combination of weapon type and bonus, and do not already have buff, then apply buff.
+			if (HasTwoHanderEquipped() && IsBuffSpell(itembonuses.WeaponStance[0])) {
+				if (!FindBuff(itembonuses.WeaponStance[0])) {
+					SpellOnTarget(itembonuses.WeaponStance[0], this);
+				}
+				weaponstance.itembonus_buff_spell_id = itembonuses.WeaponStance[0];
+			}
+
```
Comment (Akkadius) [inline zone/client.cpp:10362]: missing newline

File: zone/client.cpp [hunk 18]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+			else if (HasDuelWeaponsEquiped() && IsBuffSpell(itembonuses.WeaponStance[2])) {
+
+				if (!FindBuff(itembonuses.WeaponStance[2])) {
+					SpellOnTarget(itembonuses.WeaponStance[2], this);
+				}
+				weaponstance.itembonus_buff_spell_id = itembonuses.WeaponStance[2];
+			}
+
+			Shout("<<<<END ITEM CHECK>>>>> CURRENT SPELL ID SAVED IS [%i]", weaponstance.itembonus_buff_spell_id);
```
Comment (Akkadius) [inline zone/client.cpp:10378]: `LogSpells(`

Comment (Akkadius) [reply zone/client.cpp:10378]:
```cpp
LogSpellsDetail("[ApplyWeaponsStance] <<<<END ITEM CHECK>>>>> CURRENT SPELL ID SAVED IS [{}]", weaponstance.itembonus_buff_spell_id);
```

File: zone/client.cpp [hunk 19]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+
+			else if (HasShieldEquiped() && IsBuffSpell(itembonuses.WeaponStance[1])) {
+				if (!FindBuff(itembonuses.WeaponStance[1])) {
+					SpellOnTarget(itembonuses.WeaponStance[1], this);
+				}
+				weaponstance.itembonus_buff_spell_id = itembonuses.WeaponStance[1];
+			}
+
+			else if (HasDuelWeaponsEquiped() && IsBuffSpell(itembonuses.WeaponStance[2])) {
+
```
Comment (Akkadius) [inline zone/client.cpp:10371]: Extra newline

File: zone/client.cpp [hunk 20]
```diff
@@ -10245,3 +10245,201 @@ void Client::SetGMStatus(int newStatus) {
...
+			else if (HasShieldEquiped() && IsBuffSpell(aabonuses.WeaponStance[1])) {
+				if (!FindBuff(aabonuses.WeaponStance[1])) {
+					SpellOnTarget(aabonuses.WeaponStance[1], this);
+				}
+				weaponstance.aabonus_buff_spell_id = aabonuses.WeaponStance[1];
+			}
+
+			else if (HasDuelWeaponsEquiped() && IsBuffSpell(aabonuses.WeaponStance[2])) {
+
```
Comment (Akkadius) [inline zone/client.cpp:10428]: Newline

File: zone/spell_effects.cpp [hunk 21]
```diff
@@ -4346,6 +4347,17 @@ void Mob::BuffFadeBySlot(int slot, bool iRecalcBonuses)
 					}
 			}
 
+			case SE_Weapon_Stance:
+			{
+				/*
+				If we click off the spell buff (or fades naturally) giving us
```
Comment (Akkadius) [inline zone/spell_effects.cpp:4353]: Tab

**Review by Akkadius — APPROVED** [review]
LGTM; chatted over call. Few comments

**Comment by mackal** [comment]
> LGTM; chatted over call. Few comments

This implementation is actually wrong, I pointed it out to @KayenEQ on discord.

**Comment by KayenEQ** [comment]
mackal is referring to the way the AA Hotkey toggle is implemented. I acknowledge mackal did a good job of informing me how live accomplishes that affect. The way I wrote it is clearly not how live does it, all though it is much more straight forward. I will however likely in the future make an effort to emulate it to use the same method as live does.

**Comment by Akkadius** [comment]
> > LGTM; chatted over call. Few comments
> 
> This implementation is actually wrong, I pointed it out to @KayenEQ on discord.

My feedback is more syntax and form than it is implementation details. I look to you and others familiar with intended spell functionality to comment and review

**Comment by KayenEQ** [comment]
Hold off on merge at the moment. Decided I want to try to get the AA toggle and structure accurate.

**Review by Akkadius — APPROVED** [review]

**Comment by Akkadius** [comment]
Made tweaks, pending merge per @KayenEQ request

**Comment by KayenEQ** [comment]
Updated the AA toggle function  to be more live like. Included full custom AA now at the bottom of the PR description for testing purposes.

Also update the behavior to be more live like. If you manually click off a buff given by weapon stance it will no longer automatically reapply. You will need to move an item in your inventory to reapply it.

**Comment by KayenEQ** [comment]
Figured out some more mechanics to the AA toggle. Hold on merge while implementing.

**Comment by Akkadius** [comment]
HOLDING!

**Comment by KayenEQ** [comment]
This should be final iteration for now. Unless anyone else has comments or suggestions it should be ready for merge.

**Pull request merged**